File: Src/Icm.Core/Localization/CompositeLocalizationRepository.vb
Namespace Icm.Localization

    Public Class CompositeLocalizationRepository
        Implements ILocalizationRepository

        Private ReadOnly _subrepositories As IEnumerable(Of ILocalizationRepository)

        Public Sub New(subrepositories As IEnumerable(Of ILocalizationRepository))
            If subrepositories Is Nothing Then
                Throw New ArgumentNullException("subrepositories")
            End If

            If subrepositories.Count = 0 Then
                Throw New ArgumentException("subrepositories")
            End If

            _subrepositories = subrepositories.ToList()
        End Sub

        Public Sub New(ParamArray subrepositories() As ILocalizationRepository)
            If subrepositories Is Nothing Then
                Throw New ArgumentNullException("subrepositories")
            End If

            If subrepositories.Count = 0 Then
                Throw New ArgumentException("subrepositories")
            End If

            _subrepositories = subrepositories.ToList()
        End Sub

        Default Public ReadOnly Property ItemForCulture(lcid As Integer, key As String) As String Implements ILocalizationRepository.ItemForCulture
            Get
                For Each subrepository In _subrepositories
                    Dim result As String = subrepository.ItemForCulture(lcid, key)
                    If result IsNot Nothing Then
                        Return result
                    End If
                Next

                Return Nothing
            End Get
        End Property
    End Class

End Namespace

File: Src/Icm.Core/Localization/CompositeLocalizationRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Icm.Localization
{

	public class CompositeLocalizationRepository : ILocalizationRepository
	{


		private readonly IEnumerable<ILocalizationRepository> _subrepositories;
		public CompositeLocalizationRepository(IEnumerable<ILocalizationRepository> subrepositories)
		{
			if (subrepositories == null) {
				throw new ArgumentNullException(nameof(subrepositories));
			}

			if (!subrepositories.Any()) {
				throw new ArgumentException("msg", nameof(subrepositories));
			}

			_subrepositories = subrepositories.ToList();
		}

		public CompositeLocalizationRepository(params ILocalizationRepository[] subrepositories)
		{
			if (subrepositories == null) {
				throw new ArgumentNullException(nameof(subrepositories));
			}

			if (!subrepositories.Any()) {
				throw new ArgumentException("msg", nameof(subrepositories));
			}

			_subrepositories = subrepositories.ToList();
		}

		public string this[int lcid, string key] {
		    get
		    {
		        return _subrepositories
                    .Select(subrepository => subrepository[lcid, key])
                    .FirstOrDefault(result => result != null);
		    }
		}
	}

}


//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
